### PR TITLE
[AIR] Add `load_checkpoint` functions to Trainers

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Often, users wish to access the data saved inside Checkpoints for various purposes. We already have the logic to extract the model, preprocessor, etc. from Checkpoints inside Predictors. This PR moves that logic into a `load_checkpoint` function added to most Trainers. This allows the user to access that logic.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/train/integrations/torch/torch_trainer.py [hunk 1]
```diff
@@ -188,3 +191,32 @@ def __init__(
...
+            checkpoint: The checkpoint to load the model and
+                preprocessor from. It is expected to be from the result of a
+                ``TorchTrainer`` run.
+            model: If the checkpoint contains a model state dict, and not
+                the model itself, then the state dict will be loaded to this
+                ``model``.
+        """
+        checkpoint_dict = checkpoint.to_dict()
+        preprocessor = checkpoint_dict.get(PREPROCESSOR_KEY, None)
+        if MODEL_KEY not in checkpoint_dict:
```
Comment (amogkam) [inline python/ray/ml/train/integrations/torch/torch_trainer.py:213]: Can we move some this code (like the key validation) into superclasses, like DataParallelTrainer?

Comment (Yard1) [reply python/ray/ml/train/integrations/torch/torch_trainer.py]: Done, PTAL!

**Review by maxpumperla — APPROVED** [review]
great addition!

**Review by ericl — COMMENTED** [review]
Hmm, if this is a static method, how about decoupling it from the Trainer interface?

I.e., instead of
`LightGBMTrainer.load_checkpoint`
users could write
`ray.integrations.light_gbm.load_checkpoint`

[This is a non-blocking comment. I'm fine if others prefer attaching it to the Trainer]

**Comment by Yard1** [comment]
I think it's neater to keep it a part of the Trainers (no need to make another import, can get messy when we want to load models from different trainers in the same file), but I am happy with making it a separate function as well.

**Review by krfricke — COMMENTED** [review]
This is great.

High-level API comments:

1. I think we should add an abstract method to the base `Trainer` class
2. In that case, we should keep it attached to the trainer. Developers implementing a new trainer will then know that they'll have to implement this method as well
3. The docstrings should include the return types
4. we may want to rename the public method to `load_model_from_checkpoint`
5. We may (to be discussed) introduce (another) `load_model_and_preprocessor_from_checkpoint` method to make it very verbose what we're getting. For preprocessor there is already a utility function

**Comment by krfricke** [comment]
For 1 and 2, we should then either move it into a non-trainer method like Eric suggested, or alternatively we can raise `NotImplementedError` (and don't _require_ overwriting) if not implemented. We should only do this if we consider this to be a special case and not the general case. I think the non-trainer function would be nice, as users can see if the method is available in the module (by IDE autocompletion for the import). It's odd to have a method for almost all trainers but not for one or two, raising AttributeNotFound errors.

2 what I meant was to include a `Returns:` section in the respective docstrings

3 `load_checkpoint` is a bit weird though because you can use `trainer.load_checkpoint(checkpoint)` - for me that would imply that the checkpoint is loaded into the trainer, and some part of state is set (e.g. for continuing training or so). That's less of a problem if we move into a non-trainer function, but still, it's more loading "something" from a checkpoint, not loading the checkpoint. I would still vote for `model`, curious what other people think

**Comment by Yard1** [comment]
If we went with a separate function, how should we call it? `load_from_checkpoint`? `unpack_checkpoint`, maybe?

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 2]
```diff
@@ -434,3 +436,24 @@ def training_loop(self) -> None:
...
         }
         tune.report(**results)
+
+    @staticmethod
+    def load_checkpoint(
+        checkpoint: Checkpoint,
+    ) -> Tuple[BaseEstimator, Optional[Preprocessor]]:
+        """Load a Checkpoint from ``SklearnTrainer``.
+
+        Return the estimator and AIR preprocessor contained within.
```
Comment (maxpumperla) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:446]: we switch wording between "estimator" and "model" in doc strings. I think we should settle for one .

Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:446]: "Estimator" is sklearn parlance, it should be used specifically for that

**Comment by Yard1** [comment]
Using functions now, PTAL

**Review by amogkam — COMMENTED** [review]
@Yard1 should we also add this to the docs?

**Comment by Yard1** [comment]
@amogkam they are there already, the docs generate everything in the `ray.ml.train.integrations.X` packages

**Comment by amogkam** [comment]
@Yard1 ah that's perfect

**Review by xwjiang2010 — APPROVED** [review]
Looks great! Thanks Antoni!

File: python/ray/ml/predictors/integrations/rl/rl_predictor.py [hunk 3]
```diff
@@ -1,17 +1,11 @@
...
 import pandas as pd
-import ray.cloudpickle as cpickle
 from ray.ml import Preprocessor, Checkpoint
-from ray.ml.constants import PREPROCESSOR_KEY
 from ray.ml.predictor import Predictor, DataBatchType
-from ray.ml.train.integrations.rl.rl_trainer import (
-    RL_TRAINER_CLASS_FILE,
-    RL_CONFIG_FILE,
-)
+from ray.ml.train.integrations.rl.rl_trainer import load_checkpoint
```
Comment (amogkam) [inline python/ray/ml/predictors/integrations/rl/rl_predictor.py:8]:
```suggestion
from ray.ml.train.integrations.rl import load_checkpoint
```

**Review by amogkam — APPROVED** [review]
LGTM!

**Pull request merged**